### PR TITLE
fuzz: compressor_fuzz_test.

### PR DESCRIPTION
**Pull request by htuch**
Fuzzer for zlib compression. While the zlib project has its own fuzzer, this
fuzzer validates that the Envoy wiring around zlib makes sense and the specific
ways we configure it are safe. The fuzzer below validates a round trip
compress-decompress pair; the decompressor itself is not fuzzed beyond whatever
the compressor emits, as it exists only as a test utility today.

This PR also introduces the ability for envoy_cc_fuzz_test targets to
depend on FuzzedDataProvider. This is a high performance alternative to
libproto-mutator for structured data when the abstraction level matches,
e.g. low level binary formats. See
https://github.com/google/fuzzing/blob/master/docs/split-inputs.md#fuzzed-data-provider
for further details.

Risk level: Low (test only)
Testing: bazel run //test/common/compressor:compressor_fuzz_test_with_libfuzzer --config=asan-fuzzer

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by asraa** [comment]
Thank you so much for adding in FuzzedDataProvider!! Very excited
Looks like it covers the API for zlib we have. LGTM

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -278,4 +278,13 @@ REPOSITORY_LOCATIONS = dict(
         strip_prefix = "re2-2019-08-01",
         urls = ["https://github.com/google/re2/archive/2019-08-01.tar.gz"],
     ),
+    # Included to access FuzzedDataProvider.h. This is compiler agnostic but
+    # provided as part of the compiler-rt source distribution. We can't use the
```
[the remark this answers is not in the document]
Comment (htuch) [reply bazel/repository_locations.bzl]: We should be able to run all tests, including the fuzz tests with the test harness, on gcc still I think.

File: test/common/compressor/compressor_fuzz_test.cc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+namespace Compressor {
+namespace Fuzz {
+
+// Fuzzer for zlib compression. While the zlib project has its own fuzzer, this
+// fuzzer validates that the Envoy wiring around zlib makes sense and the
+// specific ways we configure it are safe. The fuzzer below validates a round
+// trip compress-decompress pair; the decompressor itself is not fuzzed beyond
+// whatever the compressor emits, as it exists only as a test utility today.
+DEFINE_FUZZER(const uint8_t* buf, size_t len) {
+  FuzzedDataProvider provider(buf, len);
```
Comment (mattklein123) [inline test/common/compressor/compressor_fuzz_test.cc]: Neat!

File: bazel/repository_locations.bzl (hunk 1 above)
[the remark this answers is not in the document]
Comment (lizan) [reply bazel/repository_locations.bzl]: I don't think that works today... gcc 7 doesn't even have `-fsanitize=fuzzer`.

[the remark this answers is not in the document]
Comment (htuch) [reply bazel/repository_locations.bzl]: Every fuzz test that we have today implements the LLVM expected contract https://github.com/envoyproxy/envoy/blob/master/test/fuzz/fuzz_runner.h#L63. When we build with `-fsanitize=fuzzer` etc., we are building a real fuzzer that uses this. In addition, we have a test driver that allows all fuzzers to be run in non-fuzzing mode, using the checked-in corpuses, at https://github.com/envoyproxy/envoy/blob/master/test/fuzz/main.cc. These run under `bazel test //test/...`. This latter class of fuzz tests should run under gcc.

[the remark this answers is not in the document]
Comment (lizan) [reply bazel/repository_locations.bzl]: OK that's fair, I'll explore if there's a better way to handle our clang/llvm dependencies.

**Pull request merged**